File: programs/benchmarks-alioth/python/revcomp.py
# The Computer Language Benchmarks Game
# http://benchmarksgame.alioth.debian.org/
#
# adapted for p2w
# Single-threaded version with hardcoded test sequence

# Complement table
COMPLEMENT: dict[str, str] = {
    "A": "T",
    "T": "A",
    "C": "G",
    "G": "C",
    "a": "T",
    "t": "A",
    "c": "G",
    "g": "C",
    "B": "V",
    "V": "B",
    "D": "H",
    "H": "D",
    "K": "M",
    "M": "K",
    "N": "N",
    "R": "Y",
    "Y": "R",
    "S": "S",
    "W": "W",
    "U": "A",
    "b": "V",
    "v": "B",
    "d": "H",
    "h": "D",
    "k": "M",
    "m": "K",
    "n": "N",
    "r": "Y",
    "y": "R",
    "s": "S",
    "w": "W",
    "u": "A",
}

# Sample FASTA input (normally would come from stdin)
FASTA_INPUT: str = """>ONE Homo sapiens alu
GGCCGGGCGCGGTGGCTCACGCCTGTAATCCCAGCACTTTGGGAGGCCGAGGCGGGCGGA
TCACCTGAGGTCAGGAGTTCGAGACCAGCCTGGCCAACATGGTGAAACCCCGTCTCTACT
AAAAATACAAAAATTAGCCGGGCGTGGTGGCGCGCGCCTGTAATCCCAGCTACTCGGGAG
GCTGAGGCAGGAGAATCGCTTGAACCCGGGAGGCGGAGGTTGCAGTGAGCCGAGATCGCG
CCACTGCACTCCAGCCTGGGCGACAGAGCGAGACTCCGTCTCAAAAA
>TWO IUB ambiguity codes
TAGGDHACHATCRYGMKWSYYBVNN
>THREE Homo sapiens frequency
ATGCATGCATGCATGCATGCATGCATGCATGCATGCATGCATGCATGCATGCATGCATGC
"""


def reverse_complement(sequence: str) -> str:
    """Compute reverse complement of a DNA sequence."""
    # Reverse the string using slice (O(N) with memory.copy)
    reversed_seq: str = sequence[::-1]

    # Build complement using list+join pattern (O(N) instead of O(N²))
    result_parts: list[str] = []
    i: int = 0
    while i < len(reversed_seq):
        base: str = reversed_seq[i]
        if base in COMPLEMENT:
            result_parts.append(COMPLEMENT[base])
        else:
            result_parts.append(base)
        i = i + 1

    return "".join(result_parts)


def format_sequence(seq: str, width: int) -> None:
    """Format sequence with line breaks."""
    i: int = 0
    while i < len(seq):
        end: int = i + width
        if end > len(seq):
            end = len(seq)
        # Use slicing instead of char-by-char building (O(W) vs O(W^2))
        line: str = seq[i:end]
        print(line)
        i = i + width


def process_fasta(fasta_input: str) -> None:
    """Process FASTA input and output reverse complements."""
    lines: list[str] = fasta_input.split("\n")
    header: str = ""
    sequence_parts: list[str] = []

    i: int = 0
    while i < len(lines):
        line: str = lines[i]
        if len(line) > 0:
            if line[0] == ">":
                # Output previous sequence if exists
                if len(sequence_parts) > 0:
                    print(header)
                    sequence: str = "".join(sequence_parts)
                    rc: str = reverse_complement(sequence)
                    format_sequence(rc, 60)
                    sequence_parts = []
                header = line
            else:
                sequence_parts.append(line)
        i = i + 1

    # Output last sequence
    if len(sequence_parts) > 0:
        print(header)
        sequence: str = "".join(sequence_parts)
        rc: str = reverse_complement(sequence)
        format_sequence(rc, 60)


def main() -> None:
    # Process the test input multiple times for benchmarking
    fasta_repeated: str = FASTA_INPUT * 10
    process_fasta(fasta_repeated)


main()

File: programs/benchmarks-alioth/zig/revcomp.zig
// The Computer Language Benchmarks Game
// http://benchmarksgame.alioth.debian.org/
//
// Zig port (Zig 0.15+)

const std = @import("std");
const posix = std.posix;

fn print(comptime fmt: []const u8, args: anytype) void {
    var buf: [256]u8 = undefined;
    const msg = std.fmt.bufPrint(&buf, fmt, args) catch unreachable;
    _ = posix.write(posix.STDOUT_FILENO, msg) catch {};
}

fn writeBytes(bytes: []const u8) void {
    _ = posix.write(posix.STDOUT_FILENO, bytes) catch {};
}

fn complement(c: u8) u8 {
    return switch (c) {
        'A', 'a' => 'T',
        'T', 't' => 'A',
        'C', 'c' => 'G',
        'G', 'g' => 'C',
        'B', 'b' => 'V',
        'V', 'v' => 'B',
        'D', 'd' => 'H',
        'H', 'h' => 'D',
        'K', 'k' => 'M',
        'M', 'm' => 'K',
        'N', 'n' => 'N',
        'R', 'r' => 'Y',
        'Y', 'y' => 'R',
        'S', 's' => 'S',
        'W', 'w' => 'W',
        'U', 'u' => 'A',
        else => c,
    };
}

fn reverseComplement(seq: []const u8, allocator: std.mem.Allocator) ![]u8 {
    const result = try allocator.alloc(u8, seq.len);
    for (0..seq.len) |i| {
        result[i] = complement(seq[seq.len - 1 - i]);
    }
    return result;
}

fn printFormatted(seq: []const u8, width: usize) void {
    var i: usize = 0;
    var line: [61]u8 = undefined;

    while (i < seq.len) {
        const line_len = if (i + width <= seq.len) width else seq.len - i;
        for (0..line_len) |j| {
            line[j] = seq[i + j];
        }
        line[line_len] = '\n';
        writeBytes(line[0 .. line_len + 1]);
        i += width;
    }
}

fn processEntry(header: []const u8, sequence: []const u8, allocator: std.mem.Allocator) !void {
    const rc = try reverseComplement(sequence, allocator);
    defer allocator.free(rc);

    writeBytes(header);
    writeBytes("\n");
    printFormatted(rc, 60);
}

pub fn main() !void {
    var gpa = std.heap.GeneralPurposeAllocator(.{}){};
    defer _ = gpa.deinit();
    const allocator = gpa.allocator();

    // Same test data as Python/C versions - repeated 10 times
    const entries = [_]struct { header: []const u8, sequence: []const u8 }{
        .{
            .header = ">ONE Homo sapiens alu",
            .sequence = "GGCCGGGCGCGGTGGCTCACGCCTGTAATCCCAGCACTTTGGGAGGCCGAGGCGGGCGGATCACCTGAGGTCAGGAGTTCGAGACCAGCCTGGCCAACATGGTGAAACCCCGTCTCTACTAAAAATACAAAAATTAGCCGGGCGTGGTGGCGCGCGCCTGTAATCCCAGCTACTCGGGAGGCTGAGGCAGGAGAATCGCTTGAACCCGGGAGGCGGAGGTTGCAGTGAGCCGAGATCGCGCCACTGCACTCCAGCCTGGGCGACAGAGCGAGACTCCGTCTCAAAAA",
        },
        .{
            .header = ">TWO IUB ambiguity codes",
            .sequence = "TAGGDHACHATCRYGMKWSYYBVNN",
        },
        .{
            .header = ">THREE Homo sapiens frequency",
            .sequence = "ATGCATGCATGCATGCATGCATGCATGCATGCATGCATGCATGCATGCATGCATGCATGC",
        },
    };

    // Process 10 times
    for (0..10) |_| {
        for (entries) |entry| {
            try processEntry(entry.header, entry.sequence, allocator);
        }
    }
}
